x-pack/apm-server/sampling/processor.go

```diff
@@ -0,0 +1,364 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package sampling
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/dgraph-io/badger/v2"
+	"github.com/pkg/errors"
+	"golang.org/x/sync/errgroup"
+
+	logs "github.com/elastic/apm-server/log"
+	"github.com/elastic/apm-server/model"
+	"github.com/elastic/apm-server/publish"
+	"github.com/elastic/apm-server/transform"
+	"github.com/elastic/apm-server/x-pack/apm-server/sampling/eventstorage"
+	"github.com/elastic/apm-server/x-pack/apm-server/sampling/pubsub"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+const (
+	badgerValueLogFileSize = 128 * 1024 * 1024
+)
+
+// ErrStopped is returned when calling ProcessTransformables on a stopped Processor.
+var ErrStopped = errors.New("processor is stopped")
+
+// Processor is a tail-sampling event processor.
+type Processor struct {
+	config Config
+	logger *logp.Logger
+	groups *traceGroups
+
+	storageMu sync.RWMutex
+	db        *badger.DB
+	storage   *eventstorage.ShardedReadWriter
+
+	stopMu   sync.Mutex
+	stopping chan struct{}
+	stopped  chan struct{}
+}
+
+// NewProcessor returns a new Processor, for tail-sampling trace events.
+func NewProcessor(config Config) (*Processor, error) {
+	if err := config.Validate(); err != nil {
+		return nil, errors.Wrap(err, "invalid tail-sampling config")
+	}
+
+	logger := logp.NewLogger(logs.Sampling)
+	badgerOpts := badger.DefaultOptions(config.StorageDir)
+	badgerOpts.ValueLogFileSize = badgerValueLogFileSize
+	badgerOpts.Logger = eventstorage.LogpAdaptor{Logger: logger}
+	db, err := badger.Open(badgerOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	eventCodec := eventstorage.JSONCodec{}
+	storage := eventstorage.New(db, eventCodec, config.TTL)
+	readWriter := storage.NewShardedReadWriter()
+
+	p := &Processor{
+		config:   config,
+		logger:   logger,
+		groups:   newTraceGroups(config.MaxTraceGroups, config.DefaultSampleRate, config.IngestRateDecayFactor),
+		db:       db,
+		storage:  readWriter,
+		stopping: make(chan struct{}),
+		stopped:  make(chan struct{}),
+	}
+	return p, nil
+}
+
+// ProcessTransformables processes events, writing head sampled transactions and
+// spans to storage (except where they are part of a trace that has already been
```

Comment:
>  writing head sampled transactions and spans to storage (except where they are part of a trace that has already been tail sampled)

a bit confused by this - only head-based unsampled transactions are passed through. Maybe adopt the comment here to be more clear on what happens to events that have head-based sampling information.

Reply:
We also pass through events for which we've already made/received a tail-sampling decision. i.e. when p.storage.IsTraceSampled returns no error. It's definitely a confusing message though, I'll try and clarify.